### PR TITLE
Relaxing Naming Restrictions with STRICT_PATH=false flag [GEOS-6811]

### PR DESCRIPTION
**Pull request by travislbrundage**
This is in response to a naming regression described in the following JIRA ticket: http://jira.codehaus.org/browse/GEOS-6811

The solution introduces a new Java System Property, STRICT_PATH, which if set to false, will relax the restrictions on names. The characters which are allowed with this flag set to false are:

* star (*)

* colon (:)

* comma (,)

* single quote (')

* ampersand (&)
  
* question mark (?)
  
* double quote (")
  
* less than (<)
  
* greater than (>)
  
* bar (|)

This will allow migration from geoserver 2.5.x to future versions for any catalogs which contained these characters.

### CODE REVIEW

File: doc/en/user/source/datadirectory/migrating.rst [hunk 1]
```diff
@@ -153,4 +153,33 @@ Backup files
                config.xml.2.2.x
           config.xml.2.2.x        
 
-       
\ No newline at end of file
+Migrating between GeoServer 2.5.x and 2.7.x
```
Comment (jodygarnett) [inline doc/en/user/source/datadirectory/migrating.rst]: I think we should backport this change, and change this to GeoServer 2.5.x and GeoServer 2.6.x

File: src/main/src/main/java/org/geoserver/catalog/impl/CatalogImpl.java [hunk 2]
```diff
@@ -1287,7 +1288,7 @@ public StyleInfo getStyleByName(String name) {
             // search by resource name
             String prefix = name.substring(0, colon);
             String resource = name.substring(colon + 1);
-
+
```
Comment (jodygarnett) [inline src/main/src/main/java/org/geoserver/catalog/impl/CatalogImpl.java]: these changes are spurious (white space and a dead import)

File: src/platform/src/test/java/org/geoserver/platform/resource/PathsTest.java [hunk 3]
```diff
@@ -6,6 +6,7 @@
 import java.util.Arrays;
 import java.util.List;
 
+import org.junit.Ignore;
```
Comment (jodygarnett) [inline src/platform/src/test/java/org/geoserver/platform/resource/PathsTest.java]: ignore @ignore as it is not used?

File: src/platform/src/main/java/org/geoserver/platform/resource/Paths.java [hunk 4]
```diff
@@ -102,7 +102,41 @@ public static String path(String... path) {
...
+    static final boolean STRICT_PATH = 
+            Boolean.valueOf(System.getProperty("STRICT_PATH", "true"));
+    
+    /**
+     * Pattern used to check for invalid file characters.
+     * <ul>
+     * <li> backslash
+     * </ul>
+     */
+    static final Pattern VALID = Pattern.compile("^[^\\\\]*$");
```
Comment (aaime) [inline src/platform/src/main/java/org/geoserver/platform/resource/Paths.java]: Backslash is invalid on Windows. Forward slash should be invalid on Linux? http://stackoverflow.com/questions/9847288/is-it-possible-to-use-in-a-filename

File: src/platform/src/main/java/org/geoserver/platform/resource/Paths.java [hunk 5]
```diff
@@ -102,7 +102,41 @@ public static String path(String... path) {
         }
         return toPath(names);
     }
-
+    
+    // runtime flag which, if true, throws an error for the WARN characters
+    static final boolean STRICT_PATH = 
+            Boolean.valueOf(System.getProperty("STRICT_PATH", "true"));
```
Comment (aaime) [inline src/platform/src/main/java/org/geoserver/platform/resource/Paths.java]: As discussed on the mailing list, this should be false by default since this commit fixes a regression.
But I also understand the desire to have things be more strict moving forward. 
How about making this a property in GeoServerInfo, allowing the users to edit it, and making it "true" in the release data directory, thus enforcing the more strict validation in new installations, while preserving functionality in old ones?

File: src/platform/src/main/java/org/geoserver/platform/resource/Paths.java (hunk 4 above)
Comment (travislbrundage) [inline src/platform/src/main/java/org/geoserver/platform/resource/Paths.java]: Good point; updated it to include forward slash as invalid.

File: src/platform/src/main/java/org/geoserver/platform/resource/Paths.java (hunk 5 above)
Comment (travislbrundage) [inline src/platform/src/main/java/org/geoserver/platform/resource/Paths.java]: I've set it to false per your suggestion, although I know you and Jody have had a long discussion about how we will interact with the user. Perhaps it is okay to leave it like this for now, and I can re-open discussion on the mailing list for the appropriate way to have the user make this informed choice?

Comment (aaime) [inline src/platform/src/main/java/org/geoserver/platform/resource/Paths.java]: It is not really back-portable with a "false" setting, which I guess was the main point of the discussion?

**Comment by travislbrundage** [comment]
Need to fix this - apparently the REGEX for validation is used in both the tokenized and raw version of the name, so invalidating "/" is incorrect

**Comment by travislbrundage** [comment]
@aaime Not sure how we can check forward slash to be an invalid character when it will split the file name on that character, treating it as a directory?

Either way, I am not sure that change is necessary here, as this was just meant to update the regression of disallowing certain style names.

**Comment by jodygarnett** [comment]
Yeah, we split the path up based on forward slash - so we are not going to be able to catch forward slash here.

**Comment by jodygarnett** [comment]
From GeoServer meeting:

Want this in, conversation seems to have stopped. This fix allows
GeoServer 2.5 users to update.
Action: merge this in to stable.

File: doc/en/user/source/datadirectory/migrating.rst [hunk 6]
```diff
@@ -153,4 +153,33 @@ Backup files
                config.xml.2.2.x
           config.xml.2.2.x        
 
-       
\ No newline at end of file
+Migrating between GeoServer 2.5.x and 2.6.x
+-------------------------------------------
+
+The catalog naming conventions became more strict in 2.6, invalidating certain characters within names. This is because certain protocols will not work correctly with certain characters in the name. The naming restrictions are automatically set to relaxed through the STRICT_PATH java system property variable. In order to ensure your names will work with all protocols, set this variable to true. ::
```
Comment (smithkm) [inline doc/en/user/source/datadirectory/migrating.rst]: This is a bit hard to follow.  It seems like you mean that the default is relaxed and it can be overridden with the property.  Then you say "This will allow, with warnings".  It's ambiguous whether you mean the relaxed default allows or those characters, or that enabling strict mode enables the warnings.

File: doc/en/user/source/datadirectory/migrating.rst [hunk 7]
```diff
@@ -153,4 +153,33 @@ Backup files
...
+  
+* double quote (")
+  
+* less than (<)
+  
+* greater than (>)
+  
+* bar (|)
+
+Be warned that some requests or protocols may behave unexpectedly when these characters are allowed. We recommend that you update your catalog to follow appropriate naming conventions.
```
Comment (smithkm) [inline doc/en/user/source/datadirectory/migrating.rst]: I'd include "and enable strict mode" to the recommendation, just to be clear.

File: doc/en/user/source/datadirectory/migrating.rst (hunk 6 above)
Comment (smithkm) [inline doc/en/user/source/datadirectory/migrating.rst]: Also might be worth noting that the property is only available from 2.6.3 onward.

Comment (travislbrundage) [inline doc/en/user/source/datadirectory/migrating.rst]: Yeah, my mistake. I had originally wrote this imagining that it was going to be set to true by default. I need to update that sentence.

File: doc/en/user/source/datadirectory/migrating.rst (hunk 7 above)
Comment (travislbrundage) [inline doc/en/user/source/datadirectory/migrating.rst]: Another oversight from the fact that I originally assumed it was defaulted to true. Thanks for the suggestion.

**Comment by smithkm** [comment]
The doc issues I pointed out can be dealt with later.  The only issue I see with slash is that it could be included in one of the elements passed to `toPath` which would lead to a path that would not "round trip" when split back apart; however, this issue is present regardless of this change.

So this PR looks good to me.

**Comment by travislbrundage** [comment]
I created a ticket to address the issue noted by Andrea and Kevin: [GEOS-6942](https://jira.codehaus.org/browse/GEOS-6942)

Fixed the minor doc inconsistencies.
